crates/hir_ty/src/checker/mutability.rs

```diff
@@ -0,0 +1,189 @@
+    pub(crate) fn new(
+        db: &'a dyn hir::HirMasterDatabase,
+        pod: &'a hir::Pod,
+        hir_file: hir::HirFile,
+        function: hir::Function,
+    ) -> Self {
+        Self {
+            db,
+            pod,
+            hir_file,
+            function,
+            binding_by_expr: HashMap::new(),
+            errors: Vec::new(),
+        }
+    }
+
+    pub(crate) fn check(mut self) -> Vec<TypeCheckError> {
+        let hir_file = self.pod.get_hir_file_by_function(self.function).unwrap();
+        let hir::Expr::Block(body) = self.function.body(self.db, *hir_file).unwrap() else {
+            panic!("Function body is not block");
+        };
+
+        for stmt in &body.stmts {
+            self.check_stmt(stmt);
+        }
+        if let Some(tail) = body.tail {
+            self.check_expr(tail);
+        }
+
+        self.errors
+    }
+
+    fn check_stmt(&mut self, stmt: &hir::Stmt) {
+        match stmt {
+            hir::Stmt::Expr {
+                expr,
+                has_semicolon: _,
+            } => self.check_expr(*expr),
+            hir::Stmt::Let { value, binding, .. } => {
+                self.binding_by_expr.insert(*value, *binding);
+                self.check_expr(*value);
+            }
+            hir::Stmt::Item { item } => match item {
+                hir::Item::Function(_) => (),
+                hir::Item::Module(_) => (),
+                hir::Item::UseItem(_) => (),
+            },
+        }
+    }
+
+    fn check_expr(&mut self, expr_id: hir::ExprId) {
+        let expr = expr_id.lookup(self.hir_file.db(self.db));
+        match expr {
+            hir::Expr::Block(block) => {
+                for stmt in &block.stmts {
+                    self.check_stmt(stmt);
+                }
+                if let Some(tail) = block.tail {
+                    self.check_expr(tail);
+                }
+            }
+            hir::Expr::If {
+                condition,
+                then_branch,
+                else_branch,
+            } => {
+                self.check_expr(*condition);
+                self.check_expr(*then_branch);
+                if let Some(else_branch) = else_branch {
+                    self.check_expr(*else_branch);
+                }
+            }
+            hir::Expr::Symbol(symbol) => match symbol {
+                hir::Symbol::Param { .. } => (),
+                // `VariableDef`で処理済みのため、ここでは何もしません
+                hir::Symbol::Local { .. } => (),
+                hir::Symbol::Missing { .. } => (),
+            },
+            hir::Expr::Binary { op, lhs, rhs } => {
+                self.check_expr(*lhs);
+                self.check_expr(*rhs);
+
+                let lhs = lhs.lookup(self.hir_file.db(self.db));
+                if let hir::Expr::Symbol(hir::Symbol::Local { binding, .. }) = lhs {
+                    let binding = binding.lookup(self.hir_file.db(self.db));
+                    if *op == hir::BinaryOp::Assign && !binding.mutable {
```

Comment:
**Scalability Issue**: The code checks for mutability during assignment but does not handle potential concurrency issues. This might lead to race conditions when multiple threads access and modify the same variable concurrently. <br> **Fix**: Implement locking mechanisms or use atomic operations to ensure thread safety during assignments to shared mutable variables. <br> **Code Suggestion**: 
 ```
 +use std::sync::Mutex;
 +
 +let binding = Mutex::new(binding.lookup(self.hir_file.db(self.db)));
 +if *op == hir::BinaryOp::Assign && !binding.lock().unwrap().mutable {
 +    this.errors
 +        .push(TypeCheckError::ImmutableReassignment { expr: expr_id });
 +}
 ```